In [115]:
# Libraries
from sklearn.model_selection import train_test_split
from sklearn.ensemble import IsolationForest
from imblearn.over_sampling import SMOTE
import matplotlib.pyplot as plt
from collections import Counter
import pandas as pd
import numpy as np
import os
print('OK');

OK


In [116]:
# Getting the dataset
df = pd.read_csv('/kaggle/input/creditcardfraud/creditcard.csv', sep = ',');
y = df['Class'];
X = df.drop(columns = 'Class');
print('OK');

OK


In [117]:
# SMOTE - Synthetic Minority Over-sampling Technique
sm = SMOTE(random_state = 42, sampling_strategy = 'minority');
X_smote, y_smote = sm.fit_resample(X, y);
print('OK');

OK


In [118]:
X_train, X_test, y_train, y_test = train_test_split(X_smote, y_smote, test_size = 0.2, random_state = 42);
print('OK');

OK


In [218]:
clf = IsolationForest(random_state = 0).fit(X_train[['V22','V24', 'Amount']]);
results = clf.predict(X_test[['V22','V24', 'Amount']]);
print('OK');

OK


In [219]:
accuracy = 0;
for i in range(len(results)): # 113.726
    if(results[i] == y_test.iloc[i]):
        accuracy = accuracy + 1;   
print('OK');

OK


In [220]:
print('%.2f'%((accuracy * 100)/ len(results)));

44.38


## Results with different tests
- With all atributtes from DataFrame : 37.86%
- With Time atributte: 30.04%
- With V1 atributte: 32.92%
- With V2 atributte: 33.09%
- With V3 atributte: 32.69%
- With V4 atributte: 30.07%
- With V5 atributte: 29.74%
- With V6 atributte: 38.98%
- With V7 atributte: 30.51%
- With V8 atributte: 37.09%
- With V9 atributte: 34.83%
- With V10 atributte: 29.92%
- With V11 atributte: 33.28%
- With V12 atributte: 24.20%
- With V13 atributte: 39.73%
- With V14 atributte: 19.84%
- With V15 atributte: 40.84%
- With V16 atributte: 17.83%
- With V17 atributte: 17.55%
- With V18 atributte: 23.68%
- With V19 atributte: 30.40%
- With V20 atributte: 37.87%
- With V21 atributte: 41.70%
- With V22 atributte: 43.15% <-
- With V23 atributte: 39.52%
- With V24 atributte: 42.77% <-
- With V25 atributte: 40.08%
- With V26 atributte: 42.97% <-
- With V27 atributte: 31.01%
- With V28 atributte: 32.46%
- With Amount atributte: 40.45% <-

- With some attributes together
    - 'V22','V24','V26' : 43.99%
    - 'V22','V24' : 42.82%
    - 'V22','V24', 'Amount' : 44.38%